src/libcore/ffi.rs

```diff
@@ -0,0 +1,106 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Implementation of a `va_list`.
+
+#[cfg(any(all(not(target_arch = "aarch64"), not(target_arch = "powerpc"),
+              not(target_arch = "x86_64")),
+          windows))]
+#[cfg(not(stage0))]
+/// Basic byte pointer variant of a `va_list`.
+type VaListImpl = *const i8;
+
+#[cfg(all(target_arch = "aarch64", not(windows)))]
+#[cfg(not(stage0))]
+#[repr(C)]
+#[derive(Debug)]
+/// AArch64 ABI implementation of a `va_list`. See the
+/// [Aarch64 Procedure Call Standard] for more details.
+///
+/// [AArch64 Procedure Call Standard]:
+/// http://infocenter.arm.com/help/topic/com.arm.doc.ihi0055b/IHI0055B_aapcs64.pdf
+struct VaListImpl {
+    stack: *mut (),
+    gr_top: *mut (),
+    vr_top: *mut (),
+    gr_offs: i32,
+    vr_offs: i32,
+}
+
+#[cfg(all(target_arch = "powerpc", not(windows)))]
+#[cfg(not(stage0))]
+#[repr(C)]
+#[derive(Debug)]
+/// PowerPC ABI implementation of a `va_list`.
+struct VaListImpl {
+    gpr: u8,
+    fpr: u8,
+    reserved: u16,
+    overflow_arg_area: *mut (),
+    reg_save_area: *mut (),
+}
+
+#[cfg(all(target_arch = "x86_64", not(windows)))]
+#[cfg(not(stage0))]
+#[repr(C)]
+#[derive(Debug)]
+/// x86_64 ABI implementation of a `va_list`.
+struct VaListImpl {
+    gp_offset: i32,
+    fp_offset: i32,
+    overflow_arg_area: *mut (),
+    reg_save_area: *mut (),
+}
+
+/// A wrapper for a `va_list`
+#[cfg(not(stage0))]
+#[lang = "va_list"]
+#[derive(Debug)]
+#[repr(transparent)]
+pub struct VaList<'a>(&'a mut VaListImpl);
+
+/// Trait wich limits the allowed types to be used with [VaList::arg]
+///
+/// [VaList::va_arg]: struct.VaList.html#method.arg
+pub trait VaArgSafe {}
+
+macro_rules! impl_va_arg_safe {
+    ($($t:ty),+) => {
+        $(
+            impl VaArgSafe for $t {}
+            impl VaArgSafe for *mut $t {}
+            impl VaArgSafe for *const $t {}
+        )+
+    }
+}
+
+impl_va_arg_safe!(i8, i16, i32, i64);
+impl_va_arg_safe!(u8, u16, u32, u64);
```

Comment:
This is missing `i128` and `u128`. I would also include `{u,i}size` here instead of separately below. (unless this macro invocation pattern was copied from somewhere else?)

Reply:
Updated the `{u,i}size` here, but `{u,i}128` doesn't work with the `va_arg` intrinsic, so I did not add them.